<a href="https://colab.research.google.com/github/christophermalone/stat360/blob/main/Handout13_LinearAlgebra.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Handout #13 : Multicollinearity -- Linear Algebra Calculations

<table width='100%' ><tr><td bgcolor='green'></td></tr></table>

## MN Gay Marriage Amendment

Consider the following data collected on counties from all across Minnesota.  The goal here is to investigate the relationship between the outcome from the 2012 ballot measure to restrict marriage between a man and women in the State of MN and various demographics. <table>
  <tr>
    <td width='50%'>
      <ul>
        <li>Response Variable: Percent Voted Yes for Amendment #1</li><br>
        <li>Variables under investigation (i.e. independent variables)</li>
        <ul>
          <li>% Voted Democrat</li>
          <li>% Voted Republican</li>
          <li>Population Per Square Mile</li>
          <li>Per Capita Income</li>
          <li>Average Household Size</li>
          <li>% Bachelor or higher Education</li>
          <li>% Unemployment</li>
          <li>% Population Living in Poverty</li>
          <li>% of Population: Age 0-17</li>
          <li>% of Population: Age 18-24</li>
          <li>% of Population: Age 25-44</li>
          <li>% of Population: Age 45-64</li>
          <li>% of Population: Age 65 and Over</li>
          <li>% of Population: White</li>
          <li>% of Population: African American</li>
          <li>% of Population: American Indian</li>
          <li>% of Population: Asian</li>
          <li>% of Population: Other</li>
          <li>% of Population: Of Hispanic Origin</li>
         </ul>
     </ul>
    </td>
    <td width='50%'>
    <p align='center'><img src="https://drive.google.com/uc?export=view&id=1yat57mNueWJ95pux9YstUjp2JuTc7E9t" width="50%" height="50%"></img></p>
      
  </td>
</tr>
<tr>
    <td colspan='2' width='100%'>
    The following is the actual ballot.  A vote of Yes suggests only a union of a man and a women shall be recognized as a marriage in Minnesota.
     <p align='center'><img src="https://drive.google.com/uc?export=view&id=1QpvU-viFXzqRyiKdZXs-GhLa9OnFvBRb"></img></p>


  </td>
</tr>
</table>

Data Folder: [OneDrive](https://mnscu-my.sharepoint.com/:f:/g/personal/aq7839yd_minnstate_edu/EmOQfrwxzzRBqq8PH_8qTmMBy-1qKgM11Hb8vzjs025EEA?e=wyShYs)



<table width='100%' ><tr><td bgcolor='green'></td></tr></table>




## Load tidyverse

The tidyverse R package will be used to assist with reading in the dataset into the current R session.

In [4]:
#@title Load Tidyverse
#load tidyverse package
library(tidyverse)

Warning message in system("timedatectl", intern = TRUE):
“running command 'timedatectl' had status 1”
── Attaching packages ─────────────────────────────────────── tidyverse 1.3.1 ──

✔ ggplot2 3.3.5     ✔ purrr   0.3.4
✔ tibble  3.1.5     ✔ dplyr   1.0.7
✔ tidyr   1.1.4     ✔ stringr 1.4.0
✔ readr   2.0.2     ✔ forcats 0.5.1

── Conflicts ────────────────────────────────────────── tidyverse_conflicts() ──
✖ dplyr::filter() masks stats::filter()
✖ dplyr::lag()    masks stats::lag()





---




## Load the initial data

In [5]:
# Reading data in using read_csv 
MN_Marriage_Amendment <- read_csv("http://www.StatsClass.org/stat360/Datasets/MN_Marriage_Amendment.csv")

Rows: 87 Columns: 23

── Column specification ────────────────────────────────────────────────────────
Delimiter: ","
chr  (2): State, County
dbl (21): FIPSCode, Pct_Yes, Pct_Democrat, Pct_Republican, PopPerSqMile, Per...


ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.



Taking a quick look at the initial data...

In [6]:
head(MN_Marriage_Amendment)

FIPSCode State County               Pct_Yes Pct_Democrat Pct_Republican
1 27001    MN    Aitkin County, MN    61.82   48.4         49.8          
2 27003    MN    Anoka County, MN     50.61   47.7         50.3          
3 27005    MN    Becker County, MN    62.92   41.8         56.4          
4 27007    MN    Beltrami County, MN  51.08   53.8         43.9          
5 27009    MN    Benton County, MN    55.39   41.9         55.6          
6 27011    MN    Big Stone County, MN 67.05   48.3         49.8          
  PopPerSqMile PerCapitaIncome AvgHousehold Size Pct_Bachelororhigher ⋯
1   8.9        24694           2.06              15.1                 ⋯
2 782.1        29894           2.70              25.8                 ⋯
3  24.7        25233           2.38              21.4                 ⋯
4  17.7        21753           2.56              29.1                 ⋯
5  94.2        23924           2.48              18.8                 ⋯
6  10.6        24960           2.15              16.8                 ⋯
  Pct_Age18to24 Pct_Age25to44 Pct_Age45to64 Pct_Age65andOlder Pct_RaceWhite
1  5.1          16.9          32.8          28.5              96           
2  8.1          27.8          28.4          10.8              87           
3  6.8          21.7          29.5          18.3              88           
4 14.6          22.3          25.1          13.7              75           
5 10.7          28.5          24.2          12.8              95           
6  5.6          18.4          29.8          25.4              98           
  Pct_RaceAfricanAmerican Pct_RaceAmericanIndian Pct_RaceAsian Pct_RaceOther
1 0.4                      2.4                   0.2           0.2          
2 4.0                      1.0                   4.0           1.5          
3 0.4                      8.0                   0.4           0.3          
4 0.6                     20.0                   0.7           0.3          
5 1.9                      0.4                   1.1           0.4          
6 0.2                      0.4                   0.1           0.3          
  Pct_Hispanic
1 0.9         
2 3.6         
3 1.2         
4 1.5         
5 1.6         
6 0.8



---



---



## Inverse Fail -- Silly Example

Suppose we have the following $X$ matrix in a regression setting.  Notice that the $1^{st}$ and $2^{nd}$ predictor variable are identical.  Fitting a regression model with these predictors would be **philosophically dumb** as there is no additional information gained from $2^{nd}$ predictor as it is the same as the $1^{st}$ predictor.

$$X = \begin{bmatrix}
1&6&6&2 \\
1&10&10&5 \\
1&12&12&6 \\
1&15&15&8 \\
1&18&18&10 \\
1&20&20&14 \\
\end{bmatrix} $$

In [38]:
#Setting up X matrix for silly example
x0 = c( 1, 1, 1, 1, 1, 1)
x1 = c( 6,10,12,15,18,20)
x2 = c( 6,10,12,15,18,20)
x3 = c( 2, 5, 6, 8,10,14 )
X <- cbind(x0,x1,x2,x3)
X

x0 x1 x2 x3
[1,] 1   6  6  2
[2,] 1  10 10  5
[3,] 1  12 12  6
[4,] 1  15 15  8
[5,] 1  18 18 10
[6,] 1  20 20 14

Next, compute $(X'X)$ which is needed to compute $Var(\hat{\bf{\beta}})$

In [40]:
#Getting X'X for the silly example
XpX <- t(X) %*% X
XpX

x0 x1   x2   x3 
x0  6   81   81  45
x1 81 1229 1229 714
x2 81 1229 1229 714
x3 45  714  714 425

Notice that the 2nd and 3rd row of the $(X'X)$ matrix are identical.  If this happens, the system of equations is singular. Most importantly, this is **computationally dumb** because if a system of equations is singular, an inverse cannot be obtained. 
$$(X'X) = \begin{bmatrix}
6&81&81&45 \\
81&1229&1229&714 \\
81&1229&1229&714 \\
45&714&714&425 \\
\end{bmatrix} $$

The following code in R fails as the inverse of a singular matrix cannot be obtained.

In [41]:
XpXinv <- solve(XpX, diag(4))

ERROR: ignored



---



---



## MN Marriage Amendment Data

First, let us consider the calulation of the standard errors for the model coefficients for Model #3.

Model #3:  $PctYes ∼ PctDemocrat + PctRepublican + PctAge65andOlder$

For the data considered in this example, the correlation between $PctDemocrat$ and $PctRepublican$ is very strong (negative).  

In [47]:
cor(MN_Marriage_Amendment$Pct_Democrat, MN_Marriage_Amendment$Pct_Republican)

[1] -0.9989434

Recall, the standard errors for the model coefficients are simply the diagonal elments of $Var(\hat{\bf{\beta}}) = \hat{\sigma}^{2} (X'X)^{-1}$, i.e. the variance / covariance matrix of the estimated effects from the model.

In [50]:
#Setting up the X matrix
x0 = rep(1,87)
x1=MN_Marriage_Amendment$Pct_Democrat
x2=MN_Marriage_Amendment$Pct_Republican
x3=MN_Marriage_Amendment$Pct_Age65andOlder
X <- cbind(x0,x1,x2,x3)

#Getting X'X
XpX <- t(X)%*%X
XpX


x0     x1        x2        x3      
x0   87.0   4072.60   4447.30  1541.70
x1 4072.6 194844.00 203949.61 72124.01
x2 4447.3 203949.61 231618.95 78862.83
x3 1541.7  72124.01  78862.83 28880.63

The $(X'X)$ matrix is given by the following.

$$(X'X) = \begin{bmatrix}
87.0&4072.6&4447.3&1541.7 \\
4072.6&194844.00&203949.61&72124.01 \\
4447.3&203949.61&231618.95&78862.83 \\
1541.7&72124.01&78862.83&28880.63 \\
\end{bmatrix} $$



Notice that the 2nd and 3rd row of the $(X'X)$ matrix are not identical, but the values in these two rows are fairly close to one another.  This system of equations is near singular which are likley to cause issues when computing an inverse.  

The following code will compute the variance of the estimated model parameters.  The inverse can be computed because the system of equations is not <i>exactly</i> singular; however, some would indicate that the calculations for the inverse are likley unstable.

In [49]:
#Getting X'X Inv
XpXinv <- solve(XpX, diag(4))

#Getting Var(Beta_hat)
Var_BetaHat = (4.705^2) * XpXinv
Var_BetaHat

[,1]          [,2]          [,3]          [,4]       
x0 23754.0940712 -243.92953088 -241.64251832  0.97303757
x1  -243.9295309    2.50678940    2.48060065 -0.01249051
x2  -241.6425183    2.48060065    2.45985976 -0.01253775
x3     0.9730376   -0.01249051   -0.01253775  0.01425290

The standard errors for the estimated model parameters for Model #3.

*   $Std. \, Error(\hat{\beta}_{1}) = \sqrt{2.50678940} = 1.5834$
*   $Std. \, Error(\hat{\beta}_{2}) = \sqrt{2.45985976} = 1.5685$
*   $Std. \, Error(\hat{\beta}_{3}) = \sqrt{0.1425290} = 0.1194$



Next, let us consider the calulation of the standard errors for the model coefficients for Model #1 (without $PctRepublican$).

Model #1:  $PctYes∼PctDemocrat + PctAge65andOlder$

In [45]:
#Setting up the X matrix
x0 = rep(1,87)
x1=MN_Marriage_Amendment$Pct_Democrat
x3=MN_Marriage_Amendment$Pct_Age65andOlder
X <- cbind(x0,x1,x3)

#Getting X'X
XpX <- t(X)%*%X

#Getting X'X Inv
XpXinv <- solve(XpX, diag(3))

#Getting Var(Beta_hat)
Var_BetaHat = (4.695^2) * XpXinv
Var_BetaHat

[,1]       [,2]          [,3]         
x0 16.4485181 -0.2484872162 -0.2575008217
x1 -0.2484872  0.0052505954  0.0001523078
x3 -0.2575008  0.0001523078  0.0141287428

The standard errors for the estimated model parameters for Model #1.

*    $Std. \, Error(\hat{\beta}_{1}) = \sqrt{0.0052505954} = 0.07247$
*    $Std. \, Error(\hat{\beta}_{3}) = \sqrt{0.0141287428} = 0.11887$


Notice, that the $Std \, Dev(\hat{\beta}_{1})$ from Model #1 is considerably lower than Model #3.  The reason this is the case is that Model #3 suffers from multicollinearity.


<strong><u>Moral of the Story</strong></u>: It is **philosophically** and **computational dumb** to ignore issues related to multicollinarity when building regression models.

End of Document